Automattic/pocket-casts-android
# Identify and skip/track few unknown playback sources

| 📘 Project: #261 | 📘 Continuation of: #343) |
|:---:|:---:|

> **Warning**
> The PR targets release/7.24 branch

This PR 
1. Explicitly passes `PlaybackSource` instead of setting it on the `PlaybackManager` so that we're sure of where those sources are coming from - da6367c3555639746a7cfad0bd4760ed256e684d
2. Explicitly passes `PlaybackSource` from `warningsHelper.streamingWarningDialg(...)` as suggested in https://github.com/Automattic/pocket-casts-android/pull/343#discussion_r983907170
3. Identifies a few `unknown` playback sources: `up_next`, `media_button_broadcast_search_action`, `chromecast`, `auto_play`, `auto_pause`. 
    - `auto_play`,  `auto_pause` sources are skipped from tracking.
    -  For `chromecast`, I'm only tracking the cast-connected scenario. Current screen source is tracked irrespective of whether the player is a simple player or a cast player. I can take it up separately if we decide to track it differently.

These changes might not eliminate all the unknowns. Hopefully, it is a step forward in identifying some of the missing ones. 

#### Test-1 Refactoring 
- Go through the testing steps from https://github.com/Automattic/pocket-casts-android/pull/343
- ✅ Verify that events are tracked as before

#### Test-2 Streaming Warning Dialog 
- Disable wifi and enable mobile network
- Make sure warn before using data is enabled in Profile -> Settings -> Storage & data use
- Play an episode so that `You're not on WiFi` alert is displayed
- Tap on `Stream Anyway` 
- ✅ Verify that the current screen source is tracked
- Re-enable wifi 

#### Test-3.1 Up Next
- Disable Profile -> Settings -> Play Up Next episode on tap 
- Play an episode
- Add a few episodes to the `Up Next` queue
- Long press on one of the episodes in `Up Next` queue
- ✅ Verify that the selected episode starts playing and is tracked with the source `up_next`
- Enable Profile -> Settings -> Play Up Next episode on tap 
- Tap on one of the episodes in `Up Next` queue 
- ✅ Verify that the selected episode starts playing and is tracked with the source `up_next`

#### Test-3.2 media_button_broadcast_search_action
- Close the application
- Test the search using the following terminal command
  `adb shell am start -a android.media.action.MEDIA_PLAY_FROM_SEARCH -p au.com.shiftyjelly.pocketcasts.debug --es query "The\ Daily"`
- ✅ Verify that an episode from a podcast title starting with "The\ Daily" starts playing and is tracked with the source `media_button_broadcast_search_action`

#### Test-3.3 Multiple episodes - play all
- Go to filters tab
- Select a filter having episodes
- Select options drop down in the tool bar and tap on Play all
- Tap Play x episodes from the "Play all" confirmation dialog
-  ✅ Verify that an episode starts playing and is tracked with the source `filters`

#### Test-4 Auto-play/ auto-pause skip tracking
- Add few episodes to up-next queue
- Play an episode
- Long press on the mini-player
- Tap on Mark played
- ✅ Verify that no playback event is tracked
- Expand the mini player 
- Drag the seek-bar to the end so that next episode is played
- ✅ Verify that no playback event is tracked
- While the episode is playing, background the app and play a song from another app
- Reopen the app
- ✅ Verify that the episode that was playing is paused and no playback event is tracked

# Checklist
N/A
